src/annotation/decorator_utils.ts

```diff
@@ -3,28 +3,40 @@ import * as METADATA_KEY from "../constants/metadata_keys";
 import { interfaces } from "../interfaces/interfaces";
 import { getFirstArrayDuplicate } from "../utils/js";
 
+function targetIsConstructorFunction<T = Object>(target:DecoratorTarget<T>): target is ConstructorFunction<T>{
+    return (target as any).prototype !== undefined;
+}
+
+interface ConstructorFunction<T = Object>{
+    new (...args:unknown[]): T,
+    prototype:T
+}
+
+export type DecoratorTarget<T = Object> = ConstructorFunction<T> |  T
```

Comment:
I believe that this is the best way of typing the target.
We differentiate between constructor function and prototype to prevent static properties and the prototype property is the proper way of doing so.
`_tagParameterOrProperty` requires `annotationTarget: Function` ( this cannot be typed better with generics ) and when target is a constructor function it is obviously so.  Although it is through a cast in `tagParameter` `_tagParameterOrProperty(METADATA_KEY.TAGGED, annotationTarget as ConstructorFunction, parameterIndex.toString(), metadata);` the type predicate could have been used instead.